introduce explicit mint and burn events for aex-141

marc0olo opened the pull request:
- improves the document format (headings etc.) in general
- introduces explicit `Mint` event for `mintable` extension
- introduces explicit `Burn` event for `burnable` extension
- introduces `total_supply` entrypoint in default interface
- renames `on_nft_received` to `on_aex141_received`
   - changes type `address` to `option(address)` for `from` (on minting `from` does not exist)
   - removes `to` param (most likely not needed) 
- removes the `swappable` extension
    - can be implemented individually, no need to be in a standard
- removes `from` param for `transfer`
- removes the `metadata_type` `IPFS` which doesn't really make sense as we could use `URL` for it (`ipfs://<hash>`)
- introduces `mintable_limit` extension
- introduces `transfer_to_contract` entrypoint

marc0olo committed:
introduce explicit mint and burn events for aex-141

opragmatico commented on AEXS/aex-141.md:
```diff
@@ -37,7 +37,8 @@ contract interface NFT =
 
     datatype event 
-        = Transfer(address, address, int)
+        = Mint(address, int)
+        | Transfer(address, address, int)
```
Burn to be added

marc0olo replied:
actually I think we don't have to add it here because `Burn` is a separate extension. that's why it's not included here. do you agree?

marc0olo replied:
but I have to re-think the whole document structure in that regards I guess 🤔 will have a look at AEX-9 again. but maybe it's also a bit messed up there

opragmatico replied:
I think defining it here is simpler than at the burnable extension (code > docs). The extension would continue only to refer to the `Burn` event.

marc0olo replied:
`Burn` event only relevant for `burnable` extension

opragmatico commented on AEXS/aex-141.md:
```diff
@@ -243,14 +244,31 @@ entrypoint is_approved_for_all(owner: address, operator: address) : bool
 datatype event 
-        = Transfer(address, address, int)
+        = Mint(address, int)
+        | Transfer(address, address, int)
```
Burn to be added

marc0olo replied:
same here, `Burn` event only relevant in `burnable` extension. not part of default interface

jyeshe reviewed (commented):
Looks good to me adding the `Burn` datatype. Maybe only simplify the `Burn` a bit more by removing the `from` to prevent inconsistencies when the `from` doesn't match the owner (or approved account).

marc0olo commented:
> Looks good to me adding the `Burn` datatype. Maybe only simplify the `Burn` a bit more by removing the `from` to prevent inconsistencies when the `from` doesn't match the owner (or approved account).

that's an interesting topic. depends on if we allow the approved operator to burn a specific token. my suggestion is to adapt `burn` entrypoint and add `from` param which we also have in the `transfer` entrypoint already. then `from` always has to be the owner of the NFT.

I found some issues with the document in general. will make the PR a draft again and include more updates here. want to polish it first.

marc0olo committed:
remove IPFS as metadata_type, remove from param for transfers, other minor improvements

marc0olo commented:
@jyeshe this doesn't include the `mapped_metadata` extension yet. will be a separate PR if implementation is finished and accepted.

I summarized all the changes in the description of the PR

marc0olo committed:
remove unnecessary params in NFTReceiver

jyeshe reviewed (commented):
Looking great. Liked the idea of having the `from` on the events and thank you for simplifying the methods.

A rename is needed for 'aex141_extensions' and in the `swap()` instead of 'amount' would say token ids.

marc0olo commented:
would also be cool to get feedback from @arjanvaneersel on this one 😅

marc0olo committed:
fix: aex141_extensions

marc0olo commented:
> A rename is needed for 'aex141_extensions' and in the `swap()` instead of 'amount' would say token ids.

ahh saw the mistake with `aex9_extensions` - fixed :)

regarding the `swap()` do you think it should return a list/set of token ids, right? do you mean `check_swap()` specifically here? because `swap()` doesn't return anything right now

marc0olo committed:
add total_supply entrypoint

jyeshe commented:
> > A rename is needed for 'aex141_extensions' and in the `swap()` instead of 'amount' would say token ids.
> 
> ahh saw the mistake with `aex9_extensions` - fixed :)
> 
> regarding the `swap()` do you think it should return a list/set of token ids, right? do you mean `check_swap()` specifically here? because `swap()` doesn't return anything right now

`swap` description would be changed to: Burns the whole balance of Call.caller and stores the same *token ids* in the swapped map.

marc0olo commented:
> `swap` description would be changed to: Burns the whole balance of Call.caller and stores the same _token ids_ in the swapped map.

amount of tokens for `check_swap` is fine from your perspective? set of `token_ids` could be too much given the amount of NFTs swapped 🤔

jyeshe commented:
Final one 'token contract' for 'nfts contract' (:

marc0olo committed:
improve description

marc0olo committed:
remove missed IPFS presence

marc0olo committed:
improve swapped return value to include all NFT ids

marc0olo committed:
remove missed from presence in transfer entrypoint

opragmatico commented on AEXS/aex-141.md:
```diff
@@ -372,52 +431,40 @@ stateful entrypoint check_swap(account: address) : int
+**Swap** - MUST trigger for each NFT that is swapped using the `swap` function.
 
-## NFTReceiver
+The swap event arguments should be as follows: `(owner,  token_id)`
```
A renaming to `(owner, amount)` for the `Swap(address, int)` event

marc0olo replied:
@jyeshe are you sure? we wouldn't emit sth. for the `token_id` then. in case of fungible tokens that's not a problem. but I think in case of NFTs I'd expect to know which specific NFTs have been swapped.

I first thought about a string which includes all IDs concatenated and comma seperated.

what do you think @thepiwo ?

opragmatico replied:
Because the swap might burn multiple tokens.

marc0olo replied:
Yes but for Burn or Mint we could also have custom entrypoints that perform batch operations where you'd have to emit the event X times.

So current assumption from me would be emitting the Swap event for each NFT.

If we don't care about the NFT id in the event we can leave as is and only use the "balance".

Another option would be concatenation in a string that includes all NFT ids

thepiwo replied:
how does ERC 721 do it?

marc0olo replied:
I don't think swap exists there. Need to look up.

marc0olo replied:
yes, couldn't find sth. like that. so it's up to us to decide how to deal with that event

opragmatico replied:
There is batch event on ERC1155 declared like this with `uint256[] _ids`:
`
event TransferBatch(address indexed _operator, address indexed _from, address indexed _to, uint256[] _ids, uint256[] _values);
`
This is not accepted: `datatype event = Swap(address, list(int))` and would go with `Swap(address)` because the indexer (including the Mdw) already knows which tokens the `address` owns to to remove ownership for these swapped tokens. Would follow the idea that data that is inferred tend to be more consistent than redundant/duplicated ones.

marc0olo replied:
`datatype event = Swap(address, list(int))` I know, this is why I only see two options in case we wanna include the NFT ids:
- emit for each NFT
OR
- use `string` and concatenate all IDs in a comma separated string

if we don't care about the IDs we can go with `Swap(address)`

thepiwo replied:
either we allow complete "following" of nfts via events or not at all, swap without reference to token doesn't make sense

marc0olo replied:
agree with you. actually I prefer `Swap(address, int)` like we have right now. the same applies for `Mint(address, int)`

you can have batch-minting but each mint must emit the event individually

opragmatico replied:
> either we allow complete "following" of nfts via events or not at all, swap without reference to token doesn't make sense

It informs the burner, that's enough for a indexer. It would make more sense than emitting multiple events for an event that happened only once (a single `swap`).

marc0olo replied:
that's a fair point. if you think it's sufficient I am fine with that. but I somehow like to be explicit with `Swap(address, int)` event for each swapped NFT. we already do the same for minting. the key difference is that this is a one-time event 🤔 

actually I have not such a strong opinion. I am fine with both.

opragmatico replied:
With current gas price, couldn't emit more than 2300 `Swap(address, int)` events on a method that only emits the event using `gasLimit=5817123` for a `GasLimitMax=5818080` (limited by the SDK)
http://localhost:4000/v2/txs?contract=ct_2XnZAKLw9VrKDo42Vt16KiHFWdqKubUBo34VRC5NxU8Ngi8iKj

marc0olo replied:
yeah, I thought about the gaslimit anyway in that regards. I think we cannot apply the same swap logic as in AEX-9 as a user could potentially have a lot of NFTs as you mentioned.

I think we even need to provide different swap entrypoints:
- `swap(int)` for a swap of a single NFT
- `swap(Set.set(int))` for a swap of multiple NFTs

and then keep `Swap(address, int)` event

marc0olo replied:
another option would be to keep `swap()` without any argument and return a tuple of `(int, int, bool)` which states `(amount_swapped, amount_remaining, finished_swap)`

then we would be flexible in terms of implementation.

still it remains open how to deal with the event. I still prefer `Swap(address, int)` for every single NFT.

opinions again? @jyeshe @thepiwo

opragmatico replied:
Two use cases for `swap` that might be a reason to avoid `Swap(address, int)` which is limited to 2300 events:
1. When a collection doesn't perform as expected by creators. The collection is then discontinued by burning all pre-minted tokens using `swap` (without pre-mint there would be no `swap`, just a stop on the contract).
2. When the creators want to decrease the planned supply after launch (instead of burning multiple times after minting).

Besides that multiple `Swap(address, int)` events for a single `swap` call still sound unusual for me.

Regarding the amounts arguments, as an indexer I would rely on the actual owned tokens by the swapper/burner.

marc0olo replied:
IMO `Swap` is only intended to be used for "swapping" NFTs to a new contract **OR** to another chain.

for burning NFTs `burn` should be used. **IF** single burns are not sufficient or efficient enough we would have to adjust that extension.

for reasons mentioned above I really like the flexibility we achieve by keeping `swap()` without any argument and return a tuple of `(int, int, bool)` which states `(amount_swapped, amount_remaining, finished_swap)`.

an additional reason for that is that AFTER a swap an address could theoretically receive new tokens and would potentially have to swap again.

keeping `Swap(address, int)` is just consistent in the way we do it already for minting and burning. `swap` isn't necessarily a one-time event.

maybe we should clarify the usage of `Swappable Extension`. at least this was my assumption all the time. cc @thepiwo

opragmatico replied:
I am following the definition of "Burns the whole balance of Call.caller and and stores the same amount in the swapped map". Based on this, someone else might prefer it than burning multiple times.

Either way, whatever the use case, could it be limited to 2300 event with `Swap(address, int)`? 
For the alternative of `(amount_swapped, amount_remaining, finished_swap)` could you give an example of how it would be used by an indexer? Not mentioning a trader because I believe traders rely on as much info as possible from a collection. It's important to know all tokens that were burned to keep track of both supply and rarity.

marc0olo replied:
@jyeshe I was maybe not clear enough regarding the tuple. the tuple `(amount_swapped, amount_remaining, finished_swap)` would be the return value when calling the function for the Caller to immediately know if there are still NFTs remaining and if `swap` would have to be called again. but I think we could simplify the tuple and just return `(amount_swapped, amount_remaining)`.

my assumption right now is that we might not be able to swap all the NFTs **at once** because there has to be lookups and we might run out of gas if somebody owns a huge amount of NFTs in a collection.

the event would still be `Swap(address, int)` for each single NFT.

let's assume `swap` entrypoint can handle ~ 2000 NFTs at once and the account owns 3000 NFTs. then you'd have to call `swap` twice:
1. `swap` -> returns `(2000, 1000)`
2. `swap` -> returns `(3000, 0)`

and according to the event `Swap(address, int)` -> it would be called 3000 times in total (for each NFT) 😅

thepiwo replied:
Yes, swap in AEX-9 was prepared to be able to swap to native tokens, another standard or similar.

For swap it would be crucial to also provide an example on how to swap in reality and a testcase for this, so we can be sure it works as expected.

marc0olo replied:
@thepiwo what's your opinion about the event and the proposal with returning the tuple with `(amount_swapped, amount_remaining)` ?

thepiwo replied:
I don't think amount_remaining is needed, also who cares about "amount of nfts" if I track nfts I need to know events for every single token otherwise its useless

opragmatico replied:
> @jyeshe I was maybe not clear enough regarding the tuple. the tuple `(amount_swapped, amount_remaining, finished_swap)` would be the return value when calling the function for the Caller to immediately know if there are still NFTs remaining and if `swap` would have to be called again. but I think we could simplify the tuple and just return `(amount_swapped, amount_remaining)`.
> 
> my assumption right now is that we might not be able to swap all the NFTs **at once** because there has to be lookups and we might run out of gas if somebody owns a huge amount of NFTs in a collection.
> 
> the event would still be `Swap(address, int)` for each single NFT.
> 
> let's assume `swap` entrypoint can handle ~ 2000 NFTs at once and the account owns 3000 NFTs. then you'd have to call `swap` twice:
> 
>     1. `swap` -> returns `(2000, 1000)`
> 
>     2. `swap` -> returns `(3000, 0)`
> 
> 
> and according to the event `Swap(address, int)` -> it would be called 3000 times in total (for each NFT) sweat_smile

Feel free to choose the best flavor for you, having the `address` on the event is what matters for an indexer and for some listener to call `check_swap`. Having it we make it work anyway (:

marc0olo replied:
> I don't think amount_remaining is needed

generally agree as we could check with `balance(address)` if needed. do we agree there could be scenarios where we would run out of gas, so we cannot assume that ALL NFTs can be swapped at once?

in this case I think `amount_remaining` makes sense in that regards so that calling `balance(address)` can be avoided. or at least `swap` should return a `bool` which indicates if the swap is finished or not 🤔 

`amount_swapped` can definitely be wiped out, agree

marc0olo replied:
suggesting to return `bool` now to indicate a complete swap

marc0olo replied:
I will perform some tests swapping large amount of tokens. I would recommend every early stage NFT contract to implement this as there might be a need to switch to new implementations later on

marc0olo replied:
removed swappable extension

jyeshe reviewed (commented):
One final renaming and would add the `token_id` besides `data` to the `on_nft_received`

marc0olo commented:
> One final renaming and would add the `token_id` besides `data` to the `on_nft_received`

ok I missed that one, sorry. actually this was anyway sth. I wanted to ask again. can you provide an example where this is useful? actually I removed everything mainly because `from` now doesn't make sense anymore due to explicit `Mint` event and avoiding the zero-address.

but I am curious if `token_id` and `to` could be relevant in some cases. I might add both again if you think it makes sense

jyeshe commented:
> > One final renaming and would add the `token_id` besides `data` to the `on_nft_received`
> 
> ok I missed that one, sorry. actually this was anyway sth. I wanted to ask again. can you provide an example where this is useful? actually I removed everything mainly because `from` now doesn't make sense anymore due to explicit `Mint` event and avoiding the zero-address.
> 
> but I am curious if `token_id` and `to` could be relevant in some cases. I might add both again if you think it makes sense

It's possible that when it subscribes and is notified it wants to do sth with the token using the id. On gaming a receiving contract might manage a character development.

marc0olo committed:
re-introduce token_id in on_nft_received entrypoint

marc0olo committed:
undo: Mint event in default interface

marc0olo committed:
add mint_cap to mintable extension

marc0olo committed:
description for extensions

marc0olo committed:
improve swappable extension and clarify Mint event

marc0olo committed:
return bool to indicate complete swap

marc0olo committed:
mint throws if mint_cap would exceed

marc0olo committed:
undo mint_cap

marc0olo committed:
improve swappable, return list of swapped tokens instead of amount

marc0olo commented:
@jyeshe made a small change to `check_swap()` which now returns a list of tokens instead of the swapped amount

marc0olo commented:
does anything speak against merging this? @jyeshe @thepiwo 

example NFT contracts reflecting this PR are updated here:
- https://github.com/aeternity/aex141-nft-collection-example/tree/master/contracts

jyeshe commented:
> does anything speak against merging this? @jyeshe @thepiwo 
> 
> example NFT contracts reflecting this PR are updated here:
> - https://github.com/aeternity/aex141-nft-collection-example/tree/master/contracts

As you've mentioned the current swapable adds complexity to the examples and to the standard. I would remove it for now as most collection creators don't need to read and  understand it.

marc0olo commented:
> As you've mentioned the current swapable adds complexity to the examples and to the standard. I would remove it for now as most collection creators don't need to read and understand it.

you are referring to the default collection example, right? I can and will remove it from there if we think this is always being copied and we want to avoid unnecessary usage of it.

however, do we agree now on the definition of `swappable` extension based on that example?

I would then make an explicit `swappable` in the https://github.com/aeternity/aex141-examples repository (which also has to be updated because it will be outdated when we merge this)

can we agree on that? @jyeshe @thepiwo ? =)

jyeshe commented:
> > As you've mentioned the current swapable adds complexity to the examples and to the standard. I would remove it for now as most collection creators don't need to read and understand it.
> 
> you are referring to the default collection example, right? I can and will remove it from there if we think this is always being copied and we want to avoid unnecessary usage of it.
> 
> however, do we agree now on the definition of `swappable` extension based on that example?
> 
> I would then make an explicit `swappable` in the https://github.com/aeternity/aex141-examples repository (which also has to be updated because it will be outdated when we merge this)
> 
> can we agree on that? @jyeshe @thepiwo ? =)

Default collection and since it's not possible to have a straightforward implementation, I would remove it from the standard as well. But leaving in the standard as is it is ok for the indexer.

marc0olo commented:
it is an extension to the standard. it's not there by default.

if we remove it now the whole discussion was completely useless 🙈 

let's keep the extension. if I personally would launch an NFT collection now I would definitely use `swappable` as we're very early here and things might change (again)

jyeshe commented:
After having followed some messages indicating that the discussion hasn't reached an end, it appeared it could be removed and added later or not (specially due to the continuation/stateful behavior complexity shown by the example).

But as I have mentioned before, even though I don't agree with the extension as is, I don't see it as a blocker for the Middleware (like also said before, it will require more rules for contract validation, including to follow multiple call states but  not impossible to track and signalize when the contract doesn't follow the standard).

marc0olo committed:
remove swappable extension

marc0olo commented:
@jyeshe @thepiwo after removing the swappable extension can we approve and merge the proposed changes? :)

marc0olo commented:
I updated the example accordingly and removed `swappable` from recommended contracts:
- https://github.com/aeternity/aex141-nft-collection-example/commit/7672fdd036c9ef8aae5944dfe51017e08939a6b7

jyeshe reviewed (approved):
:rocket:

marc0olo commented:
I think we have to include the `from` again to `on_nft_received` to be able to store who is the previous owner to send the tokens back to in case needed (e.g. claim the token back from a marketplace contract).

it might not be relevant in many cases and with data attribute there would be another way to handle it and provide the address of the original owner.

but providing the info explicitly can make life of devs easier 😅 

the respective contract which the token belongs to `Call.caller`. this is important to know which contract to use to handle token related actions

jyeshe commented:
True, might be useful for some liquidity market based on NFT lending.

marc0olo committed:
renaming, improving AEX141Receiver interface

marc0olo commented:
@jyeshe @thepiwo updated the example again. should reflect the state of the nft-collection example again:
- https://github.com/aeternity/aex141-nft-collection-example/tree/master/contracts
   - also added tests for:
      - approvals
      - operators
      - aex141receiver

marc0olo committed:
improve text

marc0olo commented:
https://github.com/aeternity/AEXs/pull/149 can also be checked and is based on the PR here. will be merged afterwards into master then. needs a change of the base

jyeshe commented:
> @jyeshe @thepiwo updated the example again. should reflect the state of the nft-collection example again:
> 
>     * https://github.com/aeternity/aex141-nft-collection-example/tree/master/contracts
>       
>       * also added tests for:
>         
>         * approvals
>         * operators
>         * aex141receiver

The `protected` parameter is not an argument on the `on_aex141_received` entrypoint. Would it be a generic one for ae contract methods?
https://github.com/aeternity/aex141-nft-collection-example/blob/a959d2dc4ceb5a8e1d125214d1bdd6c66d8889d4/contracts/CollectionUniqueNFTs.aes#L174
https://github.com/aeternity/aex141-nft-collection-example/blob/52bac76ff23510f3e1ddbc497fe9f5d7081a258e/contracts/recommended/CollectionTemplateEditionNFTs.aes#L215

opragmatico commented on AEXS/aex-141.md:
````diff
@@ -316,26 +316,27 @@ ApprovalForAll(address, address, string)
 
 ```sophia
-entrypoint on_nft_received(token_id: int, data: option(string)) : bool
+entrypoint on_aex141_received(from: option(address), token_id: int, data: option(string)) : bool
````
:+1:

marc0olo commented:
> The `protected` parameter is not an argument on the `on_aex141_received` entrypoint. Would it be a generic one for ae contract methods?

the protected param is a Sophia feature and an implementation detail that should not be part of the standard IMO. it prevents the contract call from reverting if there is an error at the remote contract, see https://docs.aeternity.com/aesophia/v7.0.1/sophia_features/?h=protected#protected-contract-calls

thepiwo commented on AEXS/aex-141.md:
```diff
@@ -13,21 +13,23 @@ Created: 2021-09-11
 ## Abstract
 
-A standard implementation of non-fungible tokens for the Aeternity ecosystem. The design goal of the primary interface is to be as compatible with ERC-721 as possible, so that anyone who can work with ERC-721 can work with this interface. However, where Sophia offers a better way, performance and efficiency should prevail over compatibility.
+A standard implementation of non-fungible tokens for the æternity ecosystem. The design goal of the primary interface is to be as compatible with ERC-721 as possible, so that anyone who can work with ERC-721 can work with this interface. However, where Sophia offers a better way, performance and efficiency should prevail over compatibility.
```
is this still true, maybe we should add a "how we differentiate from ERC 721" section

marc0olo replied:
it was definitely the goal and there are still some similarities. however, there are also various places we differ. so the basic NFT interface is not really compatible.

might be good to change this and introduce a section like you proposed

marc0olo replied:
updated: https://github.com/aeternity/AEXs/commit/2820512fba310a10bb3c15bce8b40c9eda016498

thepiwo commented on AEXS/aex-141.md:
```diff
@@ -13,21 +13,23 @@ Created: 2021-09-11
 
 There is no support for unsafe transactions. Therefore all transactions are ought to be safe.
 
+For minting it is expected that different options will evolve over time. The way minting is done has to be defined via extensions.
```
doesn't make sense for me to include like this

marc0olo replied:
you mean completely remove it or just re-phrase in some way? I am fine with removing it 😅

thepiwo replied:
I don't understand the meaning behind, but if this is something that needs explanation we should re-phrase

marc0olo replied:
the meaning is that there can be different types of minting processes like we already see in our recent proposal:
- regular minting
- template minting

and maybe somebody else will come up with other minting proposals in the future

marc0olo replied:
removed here: https://github.com/aeternity/AEXs/commit/2820512fba310a10bb3c15bce8b40c9eda016498

discussion about different minting events is already in another discussion to be resolved

thepiwo replied:
ok

thepiwo commented on AEXS/aex-141.md:
````diff
@@ -13,21 +13,23 @@ Created: 2021-09-11
 
 ```sophia
-contract interface NFT =
-    datatype metadata_type = URL | IPFS | OBJECT_ID | MAP
````
is ipfs now supposed to be `OBJECT_ID`?

marc0olo replied:
IPFS is now supposed to be `URL` (e.g. `ipfs://`) like we proposed in metadata of collections, too.

thepiwo commented on AEXS/aex-141.md:
```diff
@@ -105,9 +109,21 @@ entrypoint metadata(token_id: int) : option(metadata)
 
 The `metadata` type to use depends on the `metadata_type` defined in the contract:
-- for `URL`, `IPFS` and `OBJECT_ID` use `MetadataIdentifier`
+- for `URL` and `OBJECT_ID` use `MetadataIdentifier`
```
no explanation here what metadata type is supposed to be used for what content

marc0olo replied:
agree, will add

marc0olo replied:
updated: https://github.com/aeternity/AEXs/commit/07261c056b5b6003749283b21a378a0a7ba7e194

marc0olo replied:
updated again: https://github.com/aeternity/AEXs/commit/ac86407155663787d4d0685e4c68f65839b4b4b7

I think now it's good. but looking forward to your comment :)

thepiwo replied:
looks good

thepiwo commented on AEXS/aex-141.md:
```diff
@@ -141,21 +157,19 @@ entrypoint owner(token_id: int) : option(address)
+Throws if:
+- `Call.caller` is NOT the current owner or NOT approved to transfer on behalf of the owner;
+- `token_id` is NOT a valid token;
+- the invocation of `IAEX141Receiver.on_aex141_received` fails.
```
in the Example on Ethereum https://ethereum.org/en/developers/tutorials/how-to-implement-an-erc721-market/ the `openTrade` function puts an item in the escrow of a contract that has previously been approved. This will not be possible using "safe transfer" on aeternity as reentrancy is forbidden, so the `on_aex141_received` can't be called back to the originating contract

marc0olo replied:
ok now I get what you mean here. mhh 🤔 

what I currently implemented is "just" a transfer to a contract where sender is tracked and data can be provided. this could definitely be a workaround and would require markets to handle such logic based on the optional data param. see: https://github.com/aeternity/aex141-nft-collection-example/blob/master/contracts/nft-receiver-example/AEX141NFTReceiverExample.aes

how could a possible solution look like to address this or is the way mentioned above fine? feedback from @mradkov would maybe also be good here :)

thepiwo replied:
I don't understand how you would solve this?

marc0olo replied:
need to verify that with an implementation example. can you maybe summarize the exact use case where you see a problem? I mean the exact order and state of the contract calls in regards to approval and token transfer

thepiwo replied:
yes I agree

marc0olo replied:
as discussed in DM I still think it's better to avoid unsafe transfers. the typical flow is like you said that a different contract could transfer the NFT to himself and then act on behalf of the user of it which is e.g. typically done in marketplace implementations.

I came up with following example implementation to demonstrate the use of `claim_from_contract` where the nft marketplace contract claims the NFT to itself and stores data about the announced sale:
- https://github.com/aeternity/aex141-nft-simple-marketplace/blob/master/contracts/SimpleNFTMarketplaceAE.aes#L32
- https://github.com/aeternity/aex141-nft-simple-marketplace/blob/master/contracts/CollectionUniqueNFTs.aes#L102

this would be a new entrypoint we would have to propose in the default interface. if we agree on that route I will do so and we can finally proceed here :)

marc0olo replied:
solved:
- https://github.com/aeternity/AEXs/pull/148/commits/ebfe98786c5badc358fb310049d74219562d357c

also updated the examples accordingly and provided a basic marketplace example to verify:
- https://github.com/aeternity/aex141-nft-simple-marketplace/blob/master/contracts/SimpleNFTMarketplaceAE.aes#L31

thepiwo commented on AEXS/aex-141.md:
```diff
@@ -300,124 +314,111 @@ ApprovalForAll(address, address, string)
-## NFTReceiver
+**Mint**
+
+This event MUST be triggered whenever a new token is minted with the `mintable` extension.
```
if the `Mint` event usage is limited to the "mintable" extension it can't be used in indexing, every mint no matter where or how it is done MUST trigger the event in order to allow that

thepiwo replied:
maybe the Event should be part of the main standard, as without minting you can never have nfts

marc0olo replied:
we had it in the main standard. I don't mind where to place. it's exactly sth. we discussed a lot and I am not very happy having to discuss it again.

anyway, IF we put it into the main standard every mint of an NFT using `mintable_template` extension will trigger both:
- `Mint`
- `TemplateMint`

marc0olo replied:
@jyeshe can you comment on that? would like to get aligned here

marc0olo replied:
@thepiwo this is related to this comment:
- https://github.com/aeternity/AEXs/pull/148#discussion_r996699326

opragmatico replied:
> @jyeshe can you comment on that? would like to get aligned here

On the first comment about when the MUST would apply to `Mint` event I would rephrase to:
"This event MUST be triggered whenever a new token is minted during contract `init` or with the `mintable` extension."

opragmatico replied:
On the `Mint` event to be part of the standard, I think @thepiwo is refering to put it on the `datatype event` definition. For me all events could be declared there, including `Burn` and `TemplateMint`, it would be helpful to show in the code a list of all events that a standard contract might emit and one thing less to adapt. I personally prefer having the option to remove rather than to add. But if `Burn` is not included I would put at least `Mint` along with `TemplateMint` because a collection might be comprised only by semi-fungible tokens as well. 

For each kind of minting it would continue to have a different event:
- `Mint` for (unique) non-fungible tokens
- `TemplateMint` for semi-fungible tokens

marc0olo replied:
@jyeshe I think we clarified that already from our side and future coding examples will also better show how extensions differ from the basic interface. we were just not able to show that in the past in a good and re-usable way.

as for the standard definition, if an event does not belong to the core interface I wouldn't place it there. why should we?

if my interpretation of @thepiwo's statement is correct he would assume `Mint` to be emitted for **ANY** NFT, no matter if it's an NFT of a template or not. if we need to do that (for whatever reason), then we need to put it into the basic interface.

marc0olo replied:
> On the first comment about when the MUST would apply to `Mint` event I would rephrase to:
> `This event MUST be triggered whenever a new token is minted during contract `init`or with the`mintable` extension.

this means that `Mint` would (again) have to be part of the core interface. and how to handle minting on `mintable_templates` extension then? emit both?
- `Mint`
- `TemplateMint`

I am also fine with defining it this way. however, I need to adjust implementation again and deploy again etc...

would be good to finally reach consensus around this specific topic

thepiwo replied:
Just one event will be fine, as long as they are part of the standard

marc0olo replied:
is it fine then like proposed right now? let's resolve then

opragmatico replied:
It's an issue of nomenclature because things that can be sold multiple times might not be considered non-fungible. 

What I am calling semi-fungible is the token created from a template. It was devised to have analogous [ERC-1155](https://github.com/ethereum/EIPs/blob/15af728fa3b40f1aeae1476ca2b83981c684b1a6/EIPS/eip-1155.md?plain=1#L19) capability that allows for each token ID to be represented by a new configurable token type, which may have its own metadata, supply and other attributes. The difference there is that it uses a single ID for multiple supply but both are thought to represent semi-fungible tokens. 

From now on gonna refer as NFT and template NFT and putting all standard events on the data type would solve this topic.

jyeshe replied:
Also if you wanna make the mintable not an extension anymore and only put the Mint on the data type I am ok with it. It's possible to check if it is template collection by the aex141_extensions method.

opragmatico replied:
Also if you wanna make the mintable not an extension anymore and or only put the Mint on the data type it would be possible to check if it is template collection by the aex141_extensions method.

`mint` would always fail though when a collection has only template nfts.

opragmatico replied:
> Just one event will be fine, as long as they are part of the standard

Could you elaborate on this? Unify Mint and TemplateMint?

marc0olo replied:
> It's an issue of nomenclature because things that can be sold multiple times might not be considered non-fungible.

in our case **EVERY** NFT will be unique :) just wanted to point that clearly out again. we do not have any "semi-fungible" tokens. at least it's not present in any proposal yet and I absolutely want to avoid introducing it to be honest 😅 

for sure people can think it's the same token. but it will have a serial no. and you can attach individual mutable attributes to each NFT


> Also if you wanna make the mintable not an extension anymore and or only put the Mint on the data type it would be possible to check if it is template collection by the aex141_extensions method.
> 
> `mint` would always fail though when a collection has only template nfts.

I would still favor to keep it separate as it's a cleaner separation.


> Could you elaborate on this? Unify Mint and TemplateMint?

my understanding is that it's **EITHER** / **OR** as it's currently in the example implementation:
- `mintable` -> `Mint`
- `mintable_templates` -> `TemplateMint`

let's get this confirmed again by @thepiwo 😄

thepiwo replied:
I'm not sure what I'm supposed to confirm, but we can do it the way @marc0olo proposes

opragmatico replied:
> > It's an issue of nomenclature because things that can be sold multiple times might not be considered non-fungible.
> 
> in our case **EVERY** NFT will be unique :) just wanted to point that clearly out again. we do not have any "semi-fungible" tokens. at least it's not present in any proposal yet and I absolutely want to avoid introducing it to be honest 😅 
> 
> for sure people can think it's the same token. but it will have a serial no. and you can attach individual mutable attributes to each NFT
> 
> 
> > Also if you wanna make the mintable not an extension anymore and or only put the Mint on the data type it would be possible to check if it is template collection by the aex141_extensions method.
> > 
> > `mint` would always fail though when a collection has only template nfts.
> 
> I would still favor to keep it separate as it's a cleaner separation.
> 
> 
> > Could you elaborate on this? Unify Mint and TemplateMint?
> 
> my understanding is that it's **EITHER** / **OR** as it's currently in the example implementation:
> - `mintable` -> `Mint`
> - `mintable_templates` -> `TemplateMint`
> 
> let's get this confirmed again by @thepiwo 😄 

Well the whole template discussion started and ended up on code with the idea of selling multiple *copies* of the same nft including the same metadata. Arguable if they are really unique only because have different ID. But for the sake of communication I can call them also Template NFTs, no worries.

marc0olo replied:
let's postpone the metadata topic. let me resolve that conversation. let's get this PR merged :)

thepiwo commented on AEXS/aex-141.md:
```diff
@@ -300,124 +314,111 @@ ApprovalForAll(address, address, string)
+| token_id | int |
 
-Deals with receiving NFTs on behalf of a contract. Contracts MUST implement this interface to be able to receive NFTs. Mint and transfer transactions will invoke the `on_nft_received` function.
+**Burn** - MUST trigger when NFTs are burned using the `burn` function.
```
if the Burn event usage is limited to the "burn" function  it can't be used in indexing, every burn no matter where or how it is done MUST trigger the event in order to allow that

marc0olo replied:
this is just a matter of re-phrasing / clarifying, right?

marc0olo replied:
fixed: https://github.com/aeternity/AEXs/commit/121c4505c8e051fb26314a29aa58902fa79a0826

marc0olo replied:
update again due to typo and formatting: https://github.com/aeternity/AEXs/commit/f3d54075fa363d953ea5a1c742b62599471e3b88

thepiwo replied:
looks good

thepiwo commented on AEXS/aex-141.md:
```diff
@@ -282,7 +296,7 @@ Approval(address, address, int, string)
-## *ApprovalForAll*
+### *ApprovalForAll*
 
 This event MUST be triggered and emitted upon a change of operator status, including revocation of approval.
```
this description doesn't include the "All" part

marc0olo replied:
fixed: https://github.com/aeternity/AEXs/commit/7a96c18d882c83541fb1701b1a37f3777602d76a

thepiwo commented on AEXS/aex-141.md:
```diff
@@ -105,9 +109,21 @@ entrypoint metadata(token_id: int) : option(metadata)
+
 ### balance\(\)
 
 Returns the account balance of another account with address `owner` if the account has a balance. If the owner address is unknown to the contract, `None` will be returned. Using `option` type as a return value allows us to determine if the account has balance of 0, more than 0, or the account has never had balance and is still unknown to the contract.
```
balance should probably be better explained as "number of nfts owned in that collection"

marc0olo replied:
fixed: https://github.com/aeternity/AEXs/commit/e26c477d3069c36b489ab8a6cfcef94daf3b4e00

thepiwo commented on AEXS/aex-141.md:
```diff
@@ -300,124 +314,111 @@ ApprovalForAll(address, address, string)
-## Extension Swappable ("swappable")
+# Extensions
 
-### swap\(\)
```
with the removal of swap we don't expect the reference tokens to be able to be swapped any more?

marc0olo replied:
sorry, missed to answer that question.

well, we just don't define any standardized way of doing it. we can reconsider this if we finally can provide good and clean examples using the new polymorphism feature of Sophia.

people could still include that in a custom implementation but we do not expect the `Swap` event in any way right now.

as you pointed out all burns (wherever they might happen) should emit `Burn` though

thepiwo replied:
yeah, swapping will then be more complicated, but ok for now

marc0olo replied:
> yeah, swapping will then be more complicated, but ok for now

I still have the implementation in the history of the example repo. we can reconsider this as soon as we have examples making use of polymorphism to implement interfaces of extensions.

as we decided we do not want to advertise usage of `swappable` I assumed we agreed on removing it completely - at least for now

thepiwo reviewed (commented):
Typos:
 - L18 "Therefore, " comma missing
 - L97 "extention"
 - L241 "wether"

marc0olo committed:
fix metadata explanation and typos

marc0olo commented:
> Typos:
> 
> * L18 "Therefore, " comma missing
> * L97 "extention"
> * L241 "wether"

fixed: https://github.com/aeternity/AEXs/commit/07261c056b5b6003749283b21a378a0a7ba7e194

marc0olo committed:
fix: description of balance

marc0olo committed:
fix: include info about 'all' in the ApprovalForAll event description

marc0olo committed:
fix: improve description of Burn event

marc0olo committed:
improve abstract and point out important differences to ERC-721

marc0olo committed:
further improve metadata section

marc0olo committed:
typo and formatting of Burn event

marc0olo committed:
introduce transfer_to_contract

marc0olo committed:
introduce mintable_limit extension

marc0olo commented:
@jyeshe @thepiwo please one final check for following changes:
- https://github.com/aeternity/AEXs/pull/148/commits/ebfe98786c5badc358fb310049d74219562d357c -> transfer_to_contract
- https://github.com/aeternity/AEXs/pull/148/commits/02404aff0e4951803662819b8a326852e7068bfa -> mintable_limit (cc @mradkov)

marc0olo committed:
add interfaces for each extension

marc0olo committed:
fix typo

marc0olo committed:
change_token_limit -> decrease_token_limit

marc0olo commented:
@jyeshe now up to date. please check :)

marc0olo commented:
@thepiwo can we merge this one already? did you check?

thepiwo reviewed (approved):
after all the discussion I think we are fine with this as well, haven't re-checked in detail though
